rftp/file_response.go

```diff
@@ -109,16 +134,19 @@ func (f *FileResponse) write(done chan<- uint16) {
 	log.Printf("Start processing file %v\n", f.index)
 	defer func() {
 		done <- f.index
-		f.pwriter.Close()
+		// If simply Close() (or equivalently CloseWithError(nil) is called, preader
+		// never returns err=io.EOF and cmd.go hangs in io.Copy(w, req). Test case:
+		// request not-existing file.
+		f.pwriter.CloseWithError(dummyError)
+		// f.pwriter.Close()
```

Comment:
That is really weird. Do you know why it behaves like that? The solution is definitely not optimal..

Reply:
Actually I can't reproduce this. If I switch to just `f.pwriter.Close()`, I get this log:
```
2020/07/23 11:10:52 file 0 at head 11 and buffer size 0
2020/07/23 11:10:52 Finished processing file 0
2020/07/23 11:10:52 fr Read: err EOF
2020/07/23 11:10:52 File testfile received (checksum is valid)
2020/07/23 11:10:52 send abort to file writer: 0
2020/07/23 11:10:52 leaving ack writer
```
This is fine, isn't it?

Reply:
Try requesting a file that does not exist. In case the file exists, I also get the expected behavior.

Reply:
I see. But I get the same behavior for both, `Close` and `CloseWithError`. But I did some tests and I think it's a deadlock in the `Read` method, where you try to acquire a lock to set `f.Err`. But that lock has previously been acquired and not been freed by `Write`, in the metadata case of the select, when it returns early, after encountering the error status.

Reply:
Oh wow, good catch!